src/backend/tty.rs

```diff
@@ -1102,13 +1104,114 @@ impl Tty {
         match drm_compositor.render_frame::<_, _>(&mut renderer, &elements, [0.; 4]) {
             Ok(res) => {
+                if let Some(screencopy) = niri
+                    .output_state
+                    .get_mut(output)
+                    .and_then(|s| s.pending_screencopy.take())
+                {
+                    let region = screencopy.region();
+                    let damage = [Rectangle::from_loc_and_size((0, 0), region.size)];
+                    let blit: anyhow::Result<_> = if let Ok(dmabuf) =
+                        dmabuf::get_dmabuf(screencopy.buffer())
+                    {
+                        renderer.bind(dmabuf).unwrap();
+
+                        res.blit_frame_result(
+                            screencopy.region().size,
+                            Transform::Normal,
+                            output.current_scale().fractional_scale(),
+                            &mut renderer,
+                            damage,
+                            [],
+                        )
+                        .map(|sync| sync.export())
+                        .map_err(Into::into)
+                    } else {
+                        let renderer = renderer.as_gles_renderer();
+                        shm::with_buffer_contents_mut(
+                            screencopy.buffer(),
+                            |shm_buffer, shm_len, buffer_data| {
+                                let region = screencopy.region();
+                                if !(buffer_data.format == wl_shm::Format::Argb8888
+                                    && buffer_data.stride == region.size.w * 4
+                                    && buffer_data.height == region.size.h
+                                    && shm_len as i32 == buffer_data.stride * buffer_data.height)
+                                {
+                                    return Err(anyhow!("the buffer format is incorrect"));
+                                }
+
+                                renderer.with_context(|gl| {
+                                    // SAFETY: i hope this is okay
+                                    unsafe {
+                                        gl.ReadPixels(
+                                            region.loc.x,
+                                            region.loc.y,
+                                            region.size.w,
+                                            region.size.h,
+                                            gles::ffi::RGBA,
+                                            gles::ffi::UNSIGNED_BYTE,
+                                            shm_buffer.cast(),
+                                        );
+                                    }
+
+                                    // Catacomb says: Convert OpenGL's RGBA to ARGB.
+                                    // sodiboo thinks: really? this looks like RGBA -> BGRA to me?
+                                    // (is it also like,
```

Comment:
Instead of converting here, you just need to advertise the right format in protocols/screencopy.rs, and then use that same format when rendering for it.

Reply:
I don't think i can, as if i change the `wl_shm::Format` sent to the buffer, `grim` doesn't like it (which is like, one of the main consumers of this protocol). Looking at the source, it should handle a bunch of formats just fine, but on my system it only seems to work with `Argb8888`. I think it's easier to just always use the more compatible format, and we can handle converting to and from this format.

Reply:
Then you can just use the same format when rendering for screencopy.

Reply:
Interesting! Yay for eliminating all unsafe code i "wrote" except for `gl.ReadPixels`